addons/l10n_se_ocr/models/account_move.py

```diff
@@ -0,0 +1,56 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import api, fields, models, _
+from odoo.exceptions import UserError
+
+class AccountMove(models.Model):
+    _inherit = 'account.move'
+
+    def _get_invoice_reference_se_ocr2_invoice(self):
+        self.ensure_one()
+
+        number = str(self.id)
+        reference = self.env['l10n_se.luhn'].append(number)
+
+        return reference
+
+    def _get_invoice_reference_se_ocr3_invoice(self):
+        self.ensure_one()
+        number = str(self.id)
+
+        number = number + str(len(number)+2)[:1]
+        reference = self.env['l10n_se.luhn'].append(number)
+
+        return reference
+
+    def _get_invoice_reference_se_ocr4_invoice(self):
+        self.ensure_one()
+        number = str(self.id)
+
+        ocr_length = self.journal_id.invoice_ocr_length
+
+        if ocr_length < 5:
+            raise UserError(_("OCR Reference Number length need to be greater then 5. Please correct settings under invoice journal settings."))
+
+        if len(number)+1 > ocr_length:
+            raise UserError(_("OCR Reference Number length is greater than allowed. Allowed length in invoice journal setting is %s.") % str(ocr_length))
+
+        number = number.ljust(ocr_length-1, '0')
+
+
+        reference = self.env['l10n_se.luhn'].append(number)
+
+        return reference
+
+    @api.onchange('partner_id')
+    def _onchange_partner_id(self):
+        """ If Vendor Bill and Vendor OCR is set, add it. """
+        if self.partner_id and self.type == 'in_invoice' and self.partner_id.use_vendor_ocr and self.partner_id.default_vendor_ocr:
+            self.invoice_payment_ref = self.partner_id.default_vendor_ocr
```

Comment:
I would like to have a bit more information about this case: does this happen a lot that the partner uses a unique number all the time.  Because I am not sure what you want to make possible on our side then (as we miss the _partner methods (but then it could be different for every partner))

Reply:
I am not sure if Sweden is really different here.

Reply:
Yes for exampel if you pay your taxes in Sweden, it's always the same number as OCR (it's actually built on orgnumber with a few more numbers) and to the best of my knowledge all companys pay taxes... ;-) Another example is if you have an account at your supplier then the OCR number is pointing at the account, the payment for the invoice is deducted from that account. I have even seen it when I pay rent. Some loans do have the same number as well bu that depends on the bank.

Reply:
Ok did some changes in last update that should solve your concern about missing onchange method calls that isn't inplace at this moment. However with the changes you could add it at a later stage.

Reply:
Another case when you will have the same OCR number is when you have set OCR generating based on partner. I did actually take a look at my bookkeeping and there where multiple cases last year when that company used the same OCR number my guess is about 25% of all my vendor bills used a fixed OCR for that company. Not counting the 24 tax payments (two per month), 12 AGI and 12 VAT.